tensorboard/webapp/runs/store/utils_test.ts

```diff
@@ -90,5 +90,138 @@ describe('run store utils test', () => {
         });
       });
     });
+
+    describe('by regex', () => {
+      it('throws error when the regex is empty', () => {
+        let errorMessage = '';
+        try {
+          groupRuns(
+            {key: GroupByKey.REGEX, regexString: ''},
+            [
+              buildRun({id: 'eid1/alpha', name: 'foo1bar1'}),
+              buildRun({id: 'eid1/beta', name: 'foo1bar2'}),
+              buildRun({id: 'eid2/beta', name: 'foo2bar1'}),
+              buildRun({id: 'eid2/gamma', name: 'gamma'}),
+            ],
+            {
+              'eid1/alpha': 'eid1',
+              'eid1/beta': 'eid1',
+              'eid2/beta': 'eid2',
+              'eid2/gamma': 'eid2',
+            }
+          );
+        } catch (error) {
+          errorMessage = error.message;
+        }
+        expect(errorMessage).toBe('Empty regex string.');
+      });
+
+      it('throws error when the regex is invalid', () => {
+        let errorMessage = '';
+        try {
+          groupRuns(
+            {key: GroupByKey.REGEX, regexString: 'foo\\d+)bar'},
+            [
+              buildRun({id: 'eid1/alpha', name: 'foo1bar1'}),
+              buildRun({id: 'eid1/beta', name: 'foo1bar2'}),
+              buildRun({id: 'eid2/beta', name: 'foo2bar1'}),
+              buildRun({id: 'eid2/gamma', name: 'gamma'}),
+            ],
+            {
+              'eid1/alpha': 'eid1',
+              'eid1/beta': 'eid1',
+              'eid2/beta': 'eid2',
+              'eid2/gamma': 'eid2',
+            }
+          );
+        } catch (error) {
+          errorMessage = error.message;
+        }
+        expect(errorMessage).toBe('Invalid regex.');
+      });
+
+      it('groups runs by regex without capture group', () => {
+        const actual = groupRuns(
+          {key: GroupByKey.REGEX, regexString: 'foo\\d+bar'},
+          [
+            buildRun({id: 'eid1/alpha', name: 'foo1bar1'}),
+            buildRun({id: 'eid1/beta', name: 'foo1bar2'}),
+            buildRun({id: 'eid2/beta', name: 'foo2bar1'}),
+            buildRun({id: 'eid2/gamma', name: 'gamma'}),
+          ],
+          {
+            'eid1/alpha': 'eid1',
+            'eid1/beta': 'eid1',
+            'eid2/beta': 'eid2',
+            'eid2/gamma': 'eid2',
+          }
+        );
+
+        expect(actual).toEqual({
+          matches: [
```

Comment:
What would happen if my tag name magically is `matches`? I think it is better to change the interface of `groupRuns` and return list of `matches` and `nonMatches`. For example,

```ts
export interface RunGroup {
  matches: Record<string, Run[]>;
  nonMatches: Run[];
}

export function groupRuns(...): RunGroup;
```

Reply:
I think you're talking about the case when the run.name is `matches` but the run is not a match to the regex?
in this case this run would be assigned to the group where  `run.id` is the key, not `matches`
anything I missed?

but anyway I updated the return type and, although feel it's more complicated, please let me know what do you think. thanks.